File: 2024/day06/ruby/solution.rb
#!/usr/bin/env ruby

require 'set'

def parse_grid(input)
  grid = input.split("\n").map { |line| line.chars }
  rows = grid.length
  cols = grid[0].length

  # Find the starting position and direction
  start_row, start_col = nil, nil
  start_dir = nil

  grid.each_with_index do |row, r|
    row.each_with_index do |cell, c|
      if ['^', 'v', '<', '>'].include?(cell)
        start_row, start_col = r, c
        start_dir = case cell
                    when '^' then 0  # up
                    when '>' then 1  # right
                    when 'v' then 2  # down
                    when '<' then 3  # left
                    end
        # Replace guard position with empty space for simulation
        grid[r][c] = '.'
        break
      end
    end
    break if start_row
  end

  [grid, rows, cols, start_row, start_col, start_dir]
end

def simulate_patrol(grid, rows, cols, start_row, start_col, start_dir, detect_loop: false)
  # Direction vectors: up, right, down, left
  dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]

  # Track visited positions
  visited_positions = Set.new
  visited_positions.add([start_row, start_col])

  # For loop detection, track (row, col, direction) states
  visited_states = Set.new if detect_loop
  visited_states.add([start_row, start_col, start_dir]) if detect_loop

  row, col, dir = start_row, start_col, start_dir

  loop do
    # Calculate next position
    dr, dc = dirs[dir]
    next_row = row + dr
    next_col = col + dc

    # Check if next position is out of bounds
    if next_row < 0 || next_row >= rows || next_col < 0 || next_col >= cols
      # Guard leaves the area - not a loop
      return detect_loop ? false : visited_positions.size
    end

    # Check if there's an obstacle ahead
    if grid[next_row][next_col] == '#'
      # Turn right 90 degrees
      dir = (dir + 1) % 4
    else
      # Move forward
      row, col = next_row, next_col
      visited_positions.add([row, col])

      if detect_loop
        state = [row, col, dir]
        if visited_states.include?(state)
          # We've been here before with the same direction - loop detected!
          return true
        end
        visited_states.add(state)
      end
    end
  end
end

def part1(input)
  grid, rows, cols, start_row, start_col, start_dir = parse_grid(input)
  simulate_patrol(grid, rows, cols, start_row, start_col, start_dir, detect_loop: false)
end

def part2(input)
  grid, rows, cols, start_row, start_col, start_dir = parse_grid(input)

  # First, get all positions the guard visits in the original path
  # We only need to try placing obstructions on positions the guard would visit
  original_path = Set.new
  dirs = [[-1, 0], [0, 1], [1, 0], [0, -1]]

  row, col, dir = start_row, start_col, start_dir
  original_path.add([row, col])

  loop do
    dr, dc = dirs[dir]
    next_row = row + dr
    next_col = col + dc

    break if next_row < 0 || next_row >= rows || next_col < 0 || next_col >= cols

    if grid[next_row][next_col] == '#'
      dir = (dir + 1) % 4
    else
      row, col = next_row, next_col
      original_path.add([row, col])
    end
  end

  # Try placing an obstruction at each position in the original path
  # (except the starting position)
  loop_count = 0

  original_path.each do |pos_r, pos_c|
    # Skip the starting position
    next if pos_r == start_row && pos_c == start_col

    # Skip positions that already have obstacles
    next if grid[pos_r][pos_c] == '#'

    # Try placing an obstruction at this position
    grid[pos_r][pos_c] = '#'

    # Simulate and check for loop
    if simulate_patrol(grid, rows, cols, start_row, start_col, start_dir, detect_loop: true)
      loop_count += 1
    end

    # Remove the obstruction
    grid[pos_r][pos_c] = '.'
  end

  loop_count
end

# Read input from file
input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
input = File.read(input_path).strip

# Solve and output
puts "Part 1: #{part1(input)}"
puts "Part 2: #{part2(input)}"

File: 2024/day06/go/solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Position struct {
	row, col int
}

type State struct {
	pos Position
	dir Direction
}

type Direction int

const (
	Up Direction = iota
	Right
	Down
	Left
)

func main() {
	fmt.Println("Part 1:", part1())
	fmt.Println("Part 2:", part2())
}

func part1() int {
	grid, startRow, startCol, startDir := parseInput()

	// Track visited positions
	visited := make(map[Position]bool)

	// Current position and direction
	row, col := startRow, startCol
	dir := startDir

	// Mark starting position as visited
	visited[Position{row, col}] = true

	// Simulate guard movement
	for {
		// Calculate next position based on current direction
		nextRow, nextCol := getNextPosition(row, col, dir)

		// Check if guard leaves the map
		if nextRow < 0 || nextRow >= len(grid) || nextCol < 0 || nextCol >= len(grid[0]) {
			break
		}

		// Check if there's an obstacle ahead
		if grid[nextRow][nextCol] == '#' {
			// Turn right 90 degrees
			dir = turnRight(dir)
		} else {
			// Move forward
			row, col = nextRow, nextCol
			visited[Position{row, col}] = true
		}
	}

	return len(visited)
}

func parseInput() ([][]rune, int, int, Direction) {
	file, err := os.Open("../input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid [][]rune
	var startRow, startCol int
	var startDir Direction

	scanner := bufio.NewScanner(file)
	row := 0
	for scanner.Scan() {
		line := scanner.Text()
		if line == "" {
			continue
		}

		gridRow := []rune(line)

		// Find starting position
		for col, ch := range gridRow {
			switch ch {
			case '^':
				startRow, startCol = row, col
				startDir = Up
				gridRow[col] = '.' // Replace with empty space
			case 'v':
				startRow, startCol = row, col
				startDir = Down
				gridRow[col] = '.' // Replace with empty space
			case '<':
				startRow, startCol = row, col
				startDir = Left
				gridRow[col] = '.' // Replace with empty space
			case '>':
				startRow, startCol = row, col
				startDir = Right
				gridRow[col] = '.' // Replace with empty space
			}
		}

		grid = append(grid, gridRow)
		row++
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	return grid, startRow, startCol, startDir
}

func getNextPosition(row, col int, dir Direction) (int, int) {
	switch dir {
	case Up:
		return row - 1, col
	case Down:
		return row + 1, col
	case Left:
		return row, col - 1
	case Right:
		return row, col + 1
	}
	return row, col
}

func turnRight(dir Direction) Direction {
	return (dir + 1) % 4
}

func checkLoop(grid [][]rune, startRow, startCol int, startDir Direction) bool {
	rows := len(grid)
	cols := len(grid[0])

	// Track states (position, direction) to detect loops
	states := make(map[State]bool)
	row, col := startRow, startCol
	dir := startDir

	for {
		// If we've seen this state before, we're in a loop
		state := State{Position{row, col}, dir}
		if states[state] {
			return true
		}
		states[state] = true

		// Calculate next position
		nextRow, nextCol := getNextPosition(row, col, dir)

		// Check if guard would leave the map
		if nextRow < 0 || nextRow >= rows || nextCol < 0 || nextCol >= cols {
			return false
		}

		// Check if there's an obstacle ahead
		if grid[nextRow][nextCol] == '#' {
			// Turn right
			dir = turnRight(dir)
		} else {
			// Move forward
			row, col = nextRow, nextCol
		}
	}
}

func part2() int {
	grid, startRow, startCol, startDir := parseInput()
	rows := len(grid)
	cols := len(grid[0])

	count := 0

	// Try placing an obstruction at each position
	for row := 0; row < rows; row++ {
		for col := 0; col < cols; col++ {
			// Skip if there's already an obstacle or if it's the starting position
			if grid[row][col] == '#' || (row == startRow && col == startCol) {
				continue
			}

			// Place temporary obstruction
			grid[row][col] = '#'

			// Check if this creates a loop
			if checkLoop(grid, startRow, startCol, startDir) {
				count++
			}

			// Remove the obstruction
			grid[row][col] = '.'
		}
	}

	return count
}
